File: Easy/Java/Best_Time_to_Buy_and_Sell_Stock(121).java
class Solution {
    //profit on ith day is min cp uptil i-1
    public int maxProfit(int[] prices) {
        if(prices.length<=1)
            return 0;
        int cp=prices[0],profit=0;
        for(int i=1;i<prices.length;i++)
        {
            int sp=prices[i];
            if(sp-cp > profit)
                profit=sp-cp;
            else if(sp<cp)
                cp=sp;
        }
        return profit;
    }
}

File: Easy/C++/Best_Time_to_Buy_and_Sell_Stock(121).cpp
class Solution {
public:
    int maxProfit(vector<int>& a) {
        int n=a.size();
        int ans=0, mx=-2;
        for(int i=n-1;i>=0;i--)
        {
            if(a[i]>mx)
                mx=a[i];
            else
            {
                if(mx-a[i]>ans)
                    ans=mx-a[i];
            }
        }
        return ans;
    }
};